src/psyclone/parse/algorithm.py

```diff
@@ -201,14 +237,54 @@ def parse(self, alg_filename):
             # Nothing relevant found.
             raise ParseError(
                 "algorithm.py:parser:parse: Program, module, function or "
-                "subroutine not found in parse tree for file "
-                "'{0}'".format(alg_filename))
+                "subroutine not found in fparser2 parse tree.")
 
         self._unique_invoke_labels = []
         self._arg_name_to_module_name = OrderedDict()
+        # Holds variable name, type and precision information.
+        self._arg_type_defns = dict()
         invoke_calls = []
 
         for statement in walk(alg_parse_tree.content):
+            if isinstance(statement,
+                          (Type_Declaration_Stmt, Data_Component_Def_Stmt)):
+                # Capture datatype information for the variable
+                spec = statement.children[0]
+                if isinstance(spec, Declaration_Type_Spec):
+                    # This is a type declaration
+                    my_type = spec.children[1].string.lower()
+                    my_precision = None
+                elif isinstance(spec, Intrinsic_Type_Spec):
+                    # This is an intrinsic declaration
+                    my_type = spec.children[0].lower()
+                    my_precision = None
+                    if isinstance(spec.children[1], Kind_Selector):
+                        my_precision = \
+                            spec.children[1].children[1].string.lower()
+                else:
+                    raise InternalError(
+                        "Expected first child of Type_Declaration_Stmt or "
+                        "Data_Component_Def_Stmt to be Declaration_Type_Spec "
+                        "or Intrinsic_Type_Spec but found '{0}'"
+                        "".format(type(spec).__name__))
+                for decl in walk(statement.children[2], (
+                        Entity_Decl, Component_Decl)):
+                    # Determine the variables names
```

Comment:
Please could you extend this comment to say that if a variable declaration is component of a derived type, its name is stored 'as is'. This means that e.g. real :: a will clash with a derived-type definition if the latter has a component named 'a'.

(I say this as it's taken me a while to understand what's going on.)

Reply:
Added.